libcaf_net/caf/net/stream_transport.hpp

```diff
@@ -74,32 +74,33 @@ class stream_transport : public stream_transport_base<Application> {
   bool handle_read_event(endpoint_manager&) override {
     CAF_LOG_TRACE(CAF_ARG2("handle", this->handle().id));
     for (size_t reads = 0; reads < this->max_consecutive_reads_; ++reads) {
-      auto buf = this->read_buf_.data() + this->collected_;
-      size_t len = this->read_threshold_ - this->collected_;
+      auto buf = this->read_buf_.data() + collected_;
+      size_t len = read_threshold_ - collected_;
       CAF_LOG_DEBUG(CAF_ARG2("missing", len));
-      auto ret = read(this->handle_, make_span(buf, len));
+      auto num_bytes = read(this->handle_, make_span(buf, len));
+      CAF_LOG_DEBUG(CAF_ARG(len) << CAF_ARG2("handle", this->handle().id)
+                                 << CAF_ARG(num_bytes));
       // Update state.
-      if (auto num_bytes = get_if<size_t>(&ret)) {
-        CAF_LOG_DEBUG(CAF_ARG(len)
-                      << CAF_ARG(this->handle_.id) << CAF_ARG(*num_bytes));
-        this->collected_ += *num_bytes;
-        if (this->collected_ >= this->read_threshold_) {
+      if (num_bytes > 0) {
+        collected_ += num_bytes;
+        if (collected_ >= read_threshold_) {
           if (auto err = this->next_layer_.handle_data(
                 *this, make_span(this->read_buf_))) {
             CAF_LOG_ERROR("handle_data failed: " << CAF_ARG(err));
             return false;
           }
           this->prepare_next_read();
         }
-      } else {
-        auto err = get<sec>(ret);
-        if (err == sec::unavailable_or_would_block) {
-          break;
-        } else {
-          CAF_LOG_DEBUG("read failed" << CAF_ARG(err));
-          this->next_layer_.handle_error(err);
-          return false;
-        }
+      } else if (num_bytes == 0) {
+        auto err = sec::socket_disconnected;
```

Comment:
Both cases run essentially identical code on error. Wouldn't a lambda be a cleaner solution?
```
auto handle_error = [this](sec err) -> bool {
  CAF_LOG_DEBUG("read failed" << CAF_ARG(err));
  this->next_layer_.handle_error(err);
  return false;
}

...

if (num_bytes == 0) {
  return handle_error(sec::socket_disconnected);
} else if (!last_socket_error_is_temporary()) {
  return handle_error(sec::socket_operation_failed);
}
```